### ISSUE TITLE
Syntax highlighting issues

### ISSUE THREAD

**Issue by danielbecroft** [vscode-abl/vscode-abl#166]
This is a follow-on from https://github.com/vscode-abl/vscode-abl/issues/160, with some specific detection issues. Apologies if this should be raised in the [abl-tmlanguage](https://github.com/chriscamicas/abl-tmlanguage/) repository instead.

version: `v1.9.16 (pre-release)`
`LanguageServerInfo: 1.15.0-SNAPSHOT-68539a1 -- ClientInfo: Visual Studio Code 1.91.1 -- JVM: Oracle Corporation 17.0.6+9-LTS-190`

**Problem 1:** Keywords and function calls are detected as block labels, rather than keywords/parameters/function calls:

Example: `no-lock:`** is detected as block label, not a keyword when it's the last item before the colon. and there are WHERE conditions E.g.
```
// textmate scopes: keyword.other.abl source.abl
for each bCustomer
    no-lock:
end

// textmate scopes: meta.block.label.abl source.abl
for each bCustomer
    where bCustomer.CustNum gt 1
    no-lock:
end. 
```
Example: `extent(variable)` or `num-entries(variable)` is not detected as a function call with parameter when within a `do vI = 1 to ...` block

```
// textmate scope (num-entries): support.function.abl meta.function-call.abl source.abl
// textmate scope (vList): variable.other.abl meta.function.arguments.abl meta.function-call.abl source.abl
message num-entries(vList).

// textmate scope (num-entries(vList): meta.block.abl source.abl
do vI = 1 to num-entries(vList):
end.
```
(Note that the `do` block parsing is picking up the entire `num-entries(vList)` as the block label.

**Problem 2**: The second (and subsequent) buffers included in a dataset definition are detected differently:

```
define temp-table ttFoo no-undo
    field FooId as integer.

define temp-table ttBar no-undo
    field BarId as integer.

// scope ( ttFoo): storage.data.table.abl meta.define.abl source.abl
// scope ( ttBar ): entity.name.type.abl meta.define.abl source.abl
define dataset dsFoo for ttFoo, ttBar.
```

**Problem 3:** keywords like `temp-table` and `buffer` etc are not highlighted when used as parameters to procedures, or in a `buffer-copy` statement:

```
// scope (1st buffer): keyword.other.abl source.abl
// scope (2nd buffer): variable.other.abl meta.function.arguments.abl source.abl
buffer ttCustomer:buffer-copy(buffer Customer:handle).

run Foo(input buffer Customer:handle).
run Foo(input temp-table ttCustomer:handle).
```

`dataset` is correct:

```
// scope (dataset): 	keyword.other.abl meta.function.arguments.abl source.abl
vDatasetHandle:create-like(dataset dsCustomer:handle).

// scope (dataset): 	keyword.other.abl meta.function.arguments.abl source.abl
run Foo(input (dataset dsCustomer:handle)).
```
The difference with the last one is the ABL forces the use of () around the arguments, this is not valid:

```
run Foo(input dataset dsCustomer:handle).
```

**Problem 4:** `dataset` and `table` are not highlighted when used in static `parameter X for` definitions. `dataset-handle` and `table-handle` appear to be correct.

```
// scope (table): storage.data.table.abl meta.define.abl source.abl
define output parameter table for ttCustomer.

// scope (dataset): storage.data.table.abl meta.define.abl source.abl
define output parameter dataset for dsCustomer.

// scope (table-handle): meta.define.abl source.abl
define output parameter table-handle vTableHandle.

// scope (dataset-handle): 	meta.define.abl source.abl
define output parameter dataset-handle vDatasetHandle.

```

Apologies if this is the wrong repo - I can raise on the other one if required.

**Comment by PeterJudgeZA** [comment]
This repo is fine - they're both under our illustrious BDFL @gquerret  ;) 

If you can add the extension version to this report, that'd be great. You can fiond that in the ABL Language Server OUTPUT tab. It should look something like the below.

```text
[2024-07-23T15:45:43.555-0400] [INFO] vscode.rootDir: c:\Users\Peter\.vscode\extensions\riversidesoftware.openedge-abl-lsp-1.9.16-win32-x64\resources
[2024-07-23T15:45:44.168-0400] [INFO] LanguageServerInfo: 1.15.0-SNAPSHOT-68539a1 -- ClientInfo: Visual Studio Code 1.91.1 -- JVM: Eclipse Adoptium 17.0.11+9
```

**Comment by danielbecroft** [comment]
Thanks @PeterJudgeZA . I'm running v1.9.16 (pre-release), and have updated the original comment to include that information.

**Comment by danielbecroft** [comment]
Sorry, one more. If a processer contains `-NEW` in the name, then strange things happen with the highlighting:

```
&global-define MY-TABLE-NEW "table"
&global-define TEST-NEW-2 "b"

// scope (MY-TABLE-NEW): storage.type.function.abl meta.define.abl source.abl
define variable v as character initial {&MY-TABLE-NEW} no-undo.

/* here it it detected as multiple tokens:
    {&MY    source.abl
    -       source.abl
    TABLE-  source.abl
    -       source.abl
    NEW    	keyword.other.abl source.abl
    }       source.abl
*/
if {&MY-TABLE-NEW} eq "table" then
do:
end.
```

It only seems to happen with `NEW`, I've tried other keywords (`BUFFER`, `INTEGER`, etc), and they all work correctly.

**Comment by PeterJudgeZA** [comment]
> The difference with the last one is the ABL forces the use of () around the arguments, this is not valid:

Things I learned today ...

**Comment by PeterJudgeZA** [comment]
Most of this is addressed in https://github.com/chriscamicas/abl-tmlanguage/pull/37 .

The dataset definitions are in https://github.com/chriscamicas/abl-tmlanguage/pull/36 .

**Comment by gquerret** [comment]
Upgraded TextMate grammar in 1.9.18. It looks like the only remaining problem is this one:
> Problem 4: dataset and table are not highlighted when used in static parameter X for definitions. dataset-handle and table-handle appear to be correct.

Thanks a lot to our BWFL @PeterJudgeZA

**Comment by PeterJudgeZA** [comment]
I'm scared to ask what the W is for ...  :D 

Problem 4 should be fixed in PR 37. If not, then I'll look at it again.

**Comment by gquerret** [comment]
Just worker 😄

**Comment by gquerret** [comment]
> Problem 4 should be fixed in PR 37. If not, then I'll look at it again.

That's right. Then I'll close the ticket, as all issues seem to be fixed.

**Comment by danielbecroft** [comment]
Sorry @PeterJudgeZA @gquerret , I've gotten the update (v1.9.20-pre-release), and most of these issues look to be resolved. Block labels still have a slight issue - it looks like it's parsing each individual token rather than the block name as a whole when it includes a `-` character, and the block starts with one of `DO`, `REPEAT` or `TRANSACTION`. The block label tokens are correctly parsed when in the `on error undo <label>, leave <label>` clause, but there are issues in the block label itself, and when used in a `LEAVE` or `NEXT` statement inside the block (if the block name starts with `TRANSACTION` then it throws the parsing of `LEAVE` and `NEXT` off as well. 

```
// scope (DO) =>	keyword.other.abl meta.block.abl source.abl
// scope (BLOCK) => meta.block.abl source.abl
DO-BLOCK:
repeat transaction
    on error undo DO-BLOCK, leave DO-BLOCK:

    // scope (leave) => 	keyword.other.abl source.abl
    if v_condition then
        leave DO-BLOCK.
    else
        next DO-BLOCK.
end.

// scope (TRANSACTION) =>	keyword.other.abl source.abl
// scope (BLOCK) => keyword.other.abl source.abl
TRANSACTION-BLOCK:
repeat transaction
    on error undo TRANSACTION-BLOCK, leave TRANSACTION-BLOCK:
        
    // same here for `TRANSACTION-BLOCK`, but also:
    // scope (leave) => 	storage.data.table.abl source.abl
    // scope (next) =>      storage.data.table.abl source.abl
    if v_condition then
        leave TRANSACTION-BLOCK.
    else
        next TRANSACTION-BLOCK.
end.

// scope (REPEAT) =>	keyword.other.abl meta.block.abl source.abl
// scope (BLOCK) =>     meta.block.abl source.abl
REPEAT-BLOCK:
repeat on error undo REPEAT-BLOCK, leave REPEAT-BLOCK:

    // same here.
    if v_condition then
        leave REPEAT-BLOCK.
    else
        next REPEAT-BLOCK.
end.
```

These all appear to be correct:
```
// scope (MY-FIRST-BLOCK): meta.block.label.abl source.abl
MY-FIRST-BLOCK:
do:
end.

// scope (CREATE-BLOCK): 	meta.block.label.abl source.abl
CREATE-BLOCK:
do:
end. 

// scope (FOR-EACH-BLOCK): meta.block.label.abl source.abl
FOR-EACH-BLOCK:
do:
end.

// scope (FIND-BLOCK): meta.block.label.abl source.abl
FIND-BLOCK:
do:
end. 
```

Let me know if you'd prefer a separate issue raised :-)

**Comment by PeterJudgeZA** [comment]
Hm, ok.  I suspect you'll see the same for something like exclusive-lock- block (I think there's a word terminator missing in the relevant regex, but need to confirm on a computer that's not my phone). Will take a look at it soonish.

We can keep it in this issue, I have enough tickets in my life 😄

-- peter

________________________________

**Comment by danielbecroft** [comment]
> Hm, ok. I suspect you'll see the same for something like exclusive-lock- block (I think there's a word terminator missing in the relevant regex, but need to confirm on a computer that's not my phone). Will take a look at it soonish. We can keep it in this issue, I have enough tickets in my life 😄

Confirmed, the same issue does exist for `EXCLUSIVE-LOCK-BLOCK`.

**Comment by danielbecroft** [comment]
I've found a few more (okay, maybe more than a few 😄  )

### `BREAK BY` clause
```
// scope (break): 	storage.data.table.abl source.abl
for each Customer
    no-lock
    break by Customer.CustNum:

end. /* */
```

### `FUNCTION` declaration
The `function` keyword in a function declaration is not detected correctly.
```
// scope (function): variable.other.abl source.abl
function DoTheThing returns character
    (
    ):

end function.
```

### Various DATASET / TEMP-TABLE options
Some serialization options for `define temp-table` / `define dataset` statements are not detected correctly, mainly `xml-node-name` and `namespace-uri/prefix`:

```
// scope (namespace-uri): 	entity.name.type.abl meta.define.abl source.abl
// scope (namespace-prefix): 	entity.name.type.abl meta.define.abl source.abl
// scope (xml-node-name): 	entity.name.type.abl meta.define.abl source.abl
define temp-table ttFoo no-undo
        namespace-uri "foo"
        namespace-prefix "prefix"
        xml-node-name "Foo"
        serialize-name "Foo"
    // scope (xml-node-name): 	entity.name.type.abl meta.define.abl source.abl
    field FooId as integer xml-node-name "id" serialize-name "id".

define temp-table ttBar no-undo
    field FooId as integer.

// scope (namespace-uri): 	storage.data.table.abl meta.define.abl source.abl
// scope (namespace-prefix): storage.data.table.abl meta.define.abl source.abl
// scope (xml-node-name): storage.data.table.abl meta.define.abl source.abl
define dataset dsFoo
    namespace-uri "ns-uri"
    namespace-prefix "ns-prefix"
    xml-node-name "foo"
    // scope (xml-node-type): storage.data.table.abl meta.define.abl source.abl
    // scope (serialize-hidden): 	storage.data.table.abl meta.define.abl source.abl
    serialize-name "foo" xml-node-type "ELEMENT" serialize-hidden
    for ttFoo, ttBar
    // scope (foreign-key-hidden): 	storage.data.table.abl meta.define.abl source.abl
    data-relation dr01 for ttFoo, ttBar relation-fields (FooId, FooId) reposition nested foreign-key-hidden not-active
    // scope (recursive):storage.data.table.abl meta.define.abl source.abl
    data-relation dr02 for ttFoo, ttFoo relation-fields (FooId, FooId) recursive.
```


### `NOT` operator
This is potentialy going to be a controversal one. The `not` operator is being parsed/treated as a function, and all functions are assumed to have the `()` around the arguments. 

```
// scope(not): variable.other.abl source.abl
if not v_condition then
    .
// scope(not): 	support.function.abl meta.function-call.abl source.abl
else if not(v_condition) then
    .
// scope (matches): keyword.operator.source.abl source.abl
// scope (begins): keyword.operator.source.abl source.abl
// scope (ne): keyword.operator.source.abl source.abl
else if v_string matches "foo" then
    .
```

I would expect it to be detected as a `keywor.operator.source.abl`, rather than a `function` (even if there are braces there). 

### Some functions don't require `()`

The other one is things like `AVAILABLE`, `AMBIGUOUS` are functions, but they don't require the parenthesis around the argument. E.g.

```
if available bCustomerthen
    .
else if ambiguous bCustomer then
   .
```

There looks to be a few of these (`LOCKED`, `NEW` are a couple more I found). It's not clear the logic or pattern, but it seems that any function that takes in a _record_ phrase does not need the parenthesis, but if it takes any other expression, it does. E.g. `BASE64-DECODE` requires the `()`. 
Source:
- AVAILABLE function - https://documentation.progress.com/output/ua/OpenEdge_latest/pdsoe/PLUGINS_ROOT/com.openedge.pdt.langref.help/rfi1424919814832.html
- AMBIGUOUS function - https://documentation.progress.com/output/ua/OpenEdge_latest/pdsoe/PLUGINS_ROOT/com.openedge.pdt.langref.help/rfi1424919810015.html

**Comment by PeterJudgeZA** [comment]
Firstly, thanks for reporting all of this in so much detail - especially with the ABL code and the wrong scopes. It helps a lot.

The some-things-are-functions bit is at least partly due to the Progress documentation being inconsistent. Some of the grammar is generated from the OE doc and the keyword list generated by the `-zgenkwlist`  switch. There are numerous keywords that are also functions / functions that are also keywords / etc.  Thank you for finding them and reporting them - they are simple enough to fix.  `RETURN` is another one of these, that I think is handled properly now.

At the moment there's no rules to capture the "functions that don't need parens" stuff. That'll be a little more work (and probably another isue, and I'll create one when I get to looking at this).

**Comment by PeterJudgeZA** [comment]
@danielbecroft Fixes for these should be in the most recent pre-release. If they are, can you close this issue?

**Comment by danielbecroft** [comment]
Hi @PeterJudgeZA , I've had a look and they all look to be resolved now except for the dataset definition:

```
// scope (namespace-uri): 	entity.name.type.abl meta.define.abl source.abl
// scope (namespace-prefix): 	entity.name.type.abl meta.define.abl source.abl
// scope (xml-node-name): 	entity.name.type.abl meta.define.abl source.abl
define temp-table ttFoo no-undo
        namespace-uri "foo"
        namespace-prefix "prefix"
        xml-node-name "Foo"
        serialize-name "Foo"
    // scope (xml-node-name): 	entity.name.type.abl meta.define.abl source.abl
    // scope (xml-node-type):   entity.name.type.abl meta.define.abl source.abl
    field FooId as integer xml-node-name "id" xml-node-type "text" serialize-name "id".

define temp-table ttBar no-undo
    field FooId as integer.

// scope (namespace-uri): 	storage.data.table.abl meta.define.abl source.abl
// scope (namespace-prefix): storage.data.table.abl meta.define.abl source.abl
// scope (xml-node-name): storage.data.table.abl meta.define.abl source.abl
define dataset dsFoo
    namespace-uri "ns-uri"
    namespace-prefix "ns-prefix"
    xml-node-name "foo"
    // scope (xml-node-type): storage.data.table.abl meta.define.abl source.abl
    // scope (serialize-hidden): 	storage.data.table.abl meta.define.abl source.abl
    serialize-name "foo" xml-node-type "ELEMENT" serialize-hidden
    for ttFoo, ttBar
    // scope (foreign-key-hidden): 	storage.data.table.abl meta.define.abl source.abl
    data-relation dr01 for ttFoo, ttBar relation-fields (FooId, FooId) reposition nested foreign-key-hidden not-active
    // scope (recursive):storage.data.table.abl meta.define.abl source.abl
    data-relation dr02 for ttFoo, ttFoo relation-fields (FooId, FooId) recursive.
```

- In the `define temp-table` statement, `xml-node-type` is incorrect
- In the `define dataset` statement, `xml-node-type` and `serialize-hidden` are incorrect
- In the `data-relation`, `foreign-key-hidden` and `recursive` are incorrect, and `relation-fields` is coming up as a function here (not sure if it should be, as it's not a function call?)

**Comment by PeterJudgeZA** [comment]
This is resolved in grammar version 1.3.8
